### DISCUSSION TITLE
Support sub-mounted routes in openapi generation

### DISCUSSION THREAD

**Discussion by provinzkraut** [fastapi/fastapi#8849] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ ] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI, APIRouter
from fastapi.routing import Mount

app = FastAPI()
router = APIRouter()


@router.get("/")
def read_root():
    return {"Hello": "World"}


app.router.routes.append(Mount("/foo", name="foo", routes=router.routes))

print(app.openapi())
```

### Description

When running the above example, you'll notice that when using a [Mount](https://www.starlette.io/routing/#submounting-routes) with routes, those routes won't be included in the openapi schema. This is not surprising, as in `openapi.utils.get_openapi` everything that's not a subclass of `routing.APIRoute` [is ignored](https://github.com/tiangolo/fastapi/blob/22528373bba6a654323de416ad5c867cbadb81bb/fastapi/openapi/utils.py#L358).

### The solution you would like
Mounts like this make life much easier when handling bigger apps, and are an integral part of larger Starlette applications, therefore it would be great to support them in openapi generation. 

Ideally FastAPI could also support mounting routers directly, to make this feature a bit easier to use, eg something like this:

```Python
app.mount_router(router, path="/foo", name="foo"))

# equivalent to
app.router.routes.append(Mount("/foo", name="foo", routes=router.routes))
```
But I think this should be a separate issue.

### Describe alternatives you've considered

The main use of using a mount like this is organizing larger apps, and being able to address routes by a common *name* prefix.
Alternatively you could add the name prefix yourself to every route you add, but this is rather inelegant, especially seeing as Starlette provides a great feature for this. Having every route in the same namespace just doesn't seem like the best solution.

Another alternative would be to use Sub-Application, but this would result in two different openapi schemas, and comes with other drawbacks such as tedious configuration etc., when you don't actually want your sub-application to behave any differently.

**Comment by Hultner** [comment]
@provinzkraut You might have missed this but isn't this what you're looking for

```python
app.include_router(router, prefix="/foo", tags=["foo"])
```

Which as far as I can see archives exactly what you're looking for.

**Comment by provinzkraut** [comment]
@Hultner 
Thank you for your suggestion!

Unfortunately, this is not the same, as tags in FastAPI are mostly cosmetic, and the routes are still all in the same routing table, so it doesn't provide url lookups in the style of `router.url_path_for("<mount_name>:<route_name>")`. 

You could work around this by manually adding the mount name as a prefix to your route names, but this is tedious and also doesn't come with the flexibility starlette provides for routing tables. It's a powerful and integral feature of starlette, why reinvent the wheel here?

**Comment by Hultner** [comment]
> @Hultner
> Thank you for your suggestion!
> 
> Unfortunately, this is not the same, as tags in FastAPI are mostly cosmetic, and the routes are still all in the same routing table, so it doesn't provide url lookups in the style of `router.url_path_for("<mount_name>:<route_name>")`.
> 
> You could work around this by manually adding the mount name as a prefix to your route names, but this is tedious and also doesn't come with the flexibility starlette provides for routing tables. It's a powerful and integral feature of starlette, why reinvent the wheel here?

I see, I haven't used this feature of Starlette. But I guess you could add a function which iterates through all the routes at startup and patches it for you automatically then? As a viable workaround I mean. I do something like that but for aliases myself.

**Comment by alimcmaster1** [comment]
I would be a +1 for this feature!
